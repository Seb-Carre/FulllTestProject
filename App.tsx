import {
  FlatList,
  Pressable,
  SafeAreaView,
  StyleSheet,
  View,
} from 'react-native';
import React, {useRef, useState} from 'react';

import Checkbox from './components/Checkbox';
import CustomText from './components/CustomText';
import IconButton from './components/IconButton';
import SearchInput from './components/SearchInput';
import {User} from './types';
import UserCard from './components/UserCard';
import {getUsersBySearchText} from './services/api';

function App(): JSX.Element {
  const [usersSelected, setUsersSelected] = useState<User[]>([]);
  const [users, setUsers] = useState<User[]>();
  const lastCall = useRef<Number>(0);

  const AddOrDeleteUserSelected = (user: User) => {
    // Check if user is already checked
    if (checkIfUserExistsInSelectedUsers(user)) {
      // we remove the user checked and in selection
      const newList = usersSelected.filter(item => item !== user);
      setUsersSelected(newList);
    } else {
      // We add this new user in the selections
      setUsersSelected(prev => {
        return [...prev, user];
      });
    }
  };

  const deleteUsers = () => {
    // Check if the users and usersSelected lists are not empty to delete the users selected
    if (users && users.length > 0 && usersSelected.length > 0) {
      // if the user is in the selection so we don't return it
      const newList = users?.filter(user => !usersSelected.includes(user));
      // set the new list without the users selected
      setUsers(newList);
      // Reset the selection
      setUsersSelected([]);
    }
  };

  const duplicateUsers = () => {
    if (users && users.length > 0 && usersSelected.length > 0) {
      // We need to loop for creating this new array, otherwise it will create a reference of userSelected and making changes to the originals users that we don't need to duplicate, we need to create a deep copy
      // More info here => https://stackoverflow.com/a/68080733 & https://stackoverflow.com/questions/38416020/deep-copy-in-es6-using-the-spread-syntax
      // For our case, using JSON.Parse is the faster method
      const addUsers: User[] = JSON.parse(JSON.stringify(usersSelected));
      // To distict the duplication of the users, adding duplicated word is helping !
      addUsers.forEach(user => {
        user.login = `${user.login} (duplicated)`;
      });
      // We finally setting the duplicates with the current list
      setUsers(prev => {
        if (prev) {
          return [...addUsers, ...prev];
        }
      });
      setUsersSelected([]);
    }
  };

  /**
   * Function to check if a user is already in the selected ones
   * @param {User} user
   */
  const checkIfUserExistsInSelectedUsers = (user: User) => {
    return usersSelected.includes(user);
  };

  /**
   * Function to select every user
   */
  const selectAllMethod = () => {
    // if we have already selected all users, we just going unselected all users
    if (users && usersSelected.length === users.length) {
      setUsersSelected([]);
    } else {
      // simply settings all users in the selection
      if (users) {
        setUsersSelected([...users]);
      }
    }
  };

  const onChangeText = (text: string) => {
    // Every time there is a text change, the selection is removed
    setUsersSelected([]);
    // To prevent multiple requests, we are going to save the last call with a random number and making a delay
    // The last call will be taken, the others will not pass the condition
    // With it, no request each time the user is changing the text to prevent the limit rate and no bad performance.
    const randomNumber = Math.random();
    // Using a useRef to not loose the state and no need to re-render for that
    lastCall.current = randomNumber;
    // if there is a text
    if (text && text.length !== 0) {
      const timeout = setTimeout(async () => {
        // if lastCall hasn't changed, the randomNumber is actually the only or last change in text
        if (lastCall.current === randomNumber) {
          // Call API
          const usersResult = await getUsersBySearchText(text);
          // If there is a result
          if (usersResult) {
            // We set the users corresponding to the search text
            setUsers(usersResult);
          }
        }
        // Every time, you have to clear the timeout to avoid memory leaks
        clearTimeout(timeout);
      }, 1000); // The delay of waiting, 1 second here, if the user hasn't made any changes during this 1 second, the app will call the api.
    } else {
      // if no text, we clear the users from the past search
      setUsers(undefined);
    }
  };

  return (
    <SafeAreaView style={styles.container}>
      <View style={styles.container}>
        <View style={styles.containerFlex}>
          <View style={styles.titleContainer}>
            <CustomText>Github Search</CustomText>
          </View>
          <SearchInput testID="App.SearchInput" onChangeText={onChangeText} />
          <View style={styles.actionsContainer}>
            <Checkbox
              mode={
                users &&
                users.length > 0 &&
                usersSelected.length === users.length
                  ? 'multiple'
                  : 'none'
              }
              onPress={selectAllMethod}
            />
            <CustomText style={styles.numberSelectedUsers}>
              {usersSelected.length}{' '}
              <CustomText style={styles.phraseSelectedUsers}>
                elements selected
              </CustomText>
            </CustomText>
            <View style={styles.iconButtons}>
              {/* You can't use a dynamic icon */}
              <IconButton
                testID="App.CopyButton"
                iconName={require('./assets/images/copy.png')}
                onPress={duplicateUsers}
              />
              <IconButton
                testID="App.DeleteButton"
                iconName={require('./assets/images/trash.png')}
                onPress={deleteUsers}
              />
            </View>
          </View>
          <View style={styles.listWrapper}>
            {users && users.length >= 1 && (
              <FlatList
                testID='App.FlatList'
                style={styles.list}
                data={users}
                keyExtractor={(user, index) => index.toString()} // use the index for the keyExtractor, specially for the duplicated users
                renderItem={({item}) => (
                  <UserCard
                    testID="App.UserCard"
                    user={item}
                    setSelected={AddOrDeleteUserSelected}
                    selected={checkIfUserExistsInSelectedUsers(item)}
                  />
                )}></FlatList>
            )}
            {/* if the search don't give anything, this text will be visible */}
            {users !== undefined && users.length === 0 && (
              <CustomText testID="App.NoResultsFound" textType="title">
                No results found !
              </CustomText>
            )}
            {/* if the user has no text in TextInput, this text will be visible */}
            {users === undefined && (
              <View style={styles.containerFlex}>
                <CustomText testID="App.BeginSearch" textType="title">
                  Please ! Begin the search by typing the user you're looking
                  for.
                </CustomText>
              </View>
            )}
          </View>
        </View>
      </View>
    </SafeAreaView>
  );
}

const styles = StyleSheet.create({
  container: {
    padding: 10,
    flex: 1,
  },
  containerFlex: {
    flex: 1,
  },
  actionsContainer: {
    marginTop: 24,
    marginBottom: 24,
    marginLeft: 16,
    marginRight: 16,
    flexDirection: 'row',
  },
  numberSelectedUsers: {
    lineHeight: 20,
    fontWeight: 'bold',
    fontSize: 17,
    marginLeft: 10,
    flex: 1,
  },
  phraseSelectedUsers: {fontWeight: 'normal'},
  iconButtons: {
    flexDirection: 'row',
  },
  titleContainer: {
    alignItems: 'center',
    justifyContent: 'center',
    height: 64,
    backgroundColor: 'gray',
    width: '100%',
  },
  listWrapper: {
    marginLeft: 16,
    marginRight: 16,
    alignContent: 'center',
    alignItems: 'center',
    flex: 1,
  },
  list: {width: '100%'},
});

export default App;
