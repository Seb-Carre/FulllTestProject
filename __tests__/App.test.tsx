import {
  appSearchInput,
  checkbox,
  copyButton,
  deleteButton,
  flatList,
  textBeginSearch,
  textNoFoundSearch,
  userCard as userCardId,
} from '../tests/IdentifiersTest';
import {
  fakeUserResponse,
  fakeUserResponseWithMaxResults,
  fakeUserResponseWithNoData,
  headers,
  headersWithRateLimit,
  headersWithRateLimit5,
} from '../tests/mockObjects';
import {
  fireEvent,
  render,
  waitFor,
  within,
} from '@testing-library/react-native';

import {Alert} from 'react-native';
import App from '../App';
import React from 'react';
import {enableFetchMocks} from 'jest-fetch-mock';
import fetchMock from 'jest-fetch-mock';
import text from '../translate/translate';

enableFetchMocks();

beforeEach(() => {
  fetchMock.mockClear();
});

it('render the screen as snapShot expected', () => {
  const {toJSON} = render(<App />);

  expect(toJSON()).toMatchSnapshot();
});

it('render the screen, make a search and get the result', async () => {
  fetchMock.mockResponseOnce(JSON.stringify(fakeUserResponse), {
    headers: headers,
    status: 200,
  });
  const searchUser = 'sebcarre'; // the search we put in our textInput, it could be anything
  const textID = fakeUserResponse.items[0].id.toString(); // this is the ID of our user we are going to see
  const login = fakeUserResponse.items[0].login.toString(); // this is the login of our user we are going to see
  const {getByTestId, queryByText} = render(<App />);
  const input = getByTestId(appSearchInput);
  // Making the insert of the text in the inputText, this going to call our api which is mock with our fakeresponse we put
  fireEvent.changeText(input, searchUser);

  // We need to make a delay because the call api is not going to be make instantly
  await waitFor(
    () => {
      // We expect that our id and login is showed
      expect(queryByText(textID)).not.toBeNull();
      expect(queryByText(login)).not.toBeNull();
    },
    {timeout: 3000},
  );
});

it('render the screen, make a search and reset the text field', async () => {
  fetchMock.mockResponseOnce(JSON.stringify(fakeUserResponse), {
    headers: headers,
    status: 200,
  });
  const searchUser = 'sebcarre';
  const textID = fakeUserResponse.items[0].id.toString(); // this is the ID of our user we are going to see
  const login = fakeUserResponse.items[0].login.toString(); // this is the login of our user we are going to see
  const {getByTestId, queryByText, queryByTestId} = render(<App />);
  const input = getByTestId(appSearchInput);
  fireEvent.changeText(input, searchUser);

  await waitFor(
    () => {
      // We expect that our id and login are showed
      expect(queryByText(textID)).not.toBeNull();
      expect(queryByText(login)).not.toBeNull();
      expect(queryByTestId(flatList)).not.toBeNull();
    },
    {timeout: 3000},
  );

  // Now we are going to reset the textInput to remove the previous search and we are going to see if our user has gone from showing on
  fireEvent.changeText(input, '');

  await waitFor(
    () => {
      // Of course, we expect that there are no userCard showed with user infos and having our text component
      expect(queryByText(textID)).toBeNull();
      expect(queryByText(login)).toBeNull();
      expect(queryByTestId(flatList)).toBeNull();
      expect(queryByTestId(textBeginSearch)).not.toBeNull();
    },
    {timeout: 1000},
  );
});

it('render the screen, make a search and have no results', async () => {
  // In this case, the result will have no results with our make response with a empty array in items returned
  fetchMock.mockResponseOnce(JSON.stringify(fakeUserResponseWithNoData), {
    headers: headers,
    status: 200,
  });
  const searchUser = 'sebcarre';
  const {getByTestId, queryByTestId} = render(<App />);
  const input = getByTestId(appSearchInput);
  fireEvent.changeText(input, searchUser);

  await waitFor(
    () => {
      // We expect that our flatlist is not showed but only our text component to say that there are no results
      expect(queryByTestId(flatList)).toBeNull();
      expect(queryByTestId(textNoFoundSearch)).not.toBeNull();
      expect(queryByTestId(textBeginSearch)).toBeNull();
    },
    {timeout: 2000},
  );
});

it('render the screen, mock api call headers with 5 calls allowed to check the alert dialog with', async () => {
  // Same thing like we have in our api unit test but with the integration of our functionnal component
  fetchMock.mockResponseOnce(JSON.stringify(fakeUserResponseWithNoData), {
    headers: headersWithRateLimit5,
    status: 200,
  });
  jest.spyOn(Alert, 'alert');
  const searchUser = 'sebcarre';
  const {getByTestId} = render(<App />);
  const input = getByTestId(appSearchInput);
  fireEvent.changeText(input, searchUser);

  await waitFor(
    () => {
      expect(Alert.alert).toHaveBeenCalledWith(
        text('warning'),
        text('alertFiveSearchsRemaining'),
      );
    },
    {timeout: 2000},
  );
});

it('render the screen, mock api call headers with no more calls allowed and status 403 to check the alert dialog with', async () => {
  // Same thing like we have in our api unit test but with the integration of our functionnal component
  fetchMock.mockResponseOnce(JSON.stringify(fakeUserResponseWithNoData), {
    headers: headersWithRateLimit,
    status: 403,
  });
  jest.spyOn(Alert, 'alert');
  const searchUser = 'sebcarre';
  const {getByTestId} = render(<App />);
  const input = getByTestId(appSearchInput);
  fireEvent.changeText(input, searchUser);
  const time = headersWithRateLimit.get('x-ratelimit-reset');
  const date = new Date(Number(time) * 1000);
  await waitFor(
    () => {
      expect(Alert.alert).toHaveBeenCalledWith(
        text('error'),
        `${text('alertNoMoreSearchsRemainingWithDate')} ${date}`,
      );
    },
    {timeout: 2000},
  );
});

it('render the screen, make a search, select a user and duplicate', async () => {
  fetchMock.mockResponseOnce(JSON.stringify(fakeUserResponse), {
    headers: headers,
    status: 200,
  });
  const searchUser = 'sebcarre';
  // This is the user we going to have with the original
  const resultUser = `${fakeUserResponse.items[0].login} ${text(
    'duplicatedUser',
  )}`;
  const textID = fakeUserResponse.items[0].id.toString();
  const {getByTestId, queryByText} = render(<App />);
  const input = getByTestId(appSearchInput);
  fireEvent.changeText(input, searchUser);

  await waitFor(
    () => {
      expect(queryByText(textID)).not.toBeNull();
      const userCard = getByTestId(userCardId + textID); // we select the right card item
      fireEvent.press(
        within(userCard).getByTestId('UserCard.CheckBox' + textID),
      );
      fireEvent.press(getByTestId(copyButton));
    },
    {timeout: 3000},
  );

  await waitFor(
    () => {
      // We expect original and duplicated to be present
      expect(queryByText(resultUser)).not.toBeNull();
      expect(queryByText(fakeUserResponse.items[0].login)).not.toBeNull();
    },
    {timeout: 1000},
  );
});

it('render the screen, make a search, select a user and remove', async () => {
  fetchMock.mockResponseOnce(JSON.stringify(fakeUserResponse), {
    headers: headers,
    status: 200,
  });
  const searchUser = 'sebcarre';
  const resultUser = fakeUserResponse.items[0].login;
  const textID = fakeUserResponse.items[0].id.toString();
  const {getByTestId, queryByText} = render(<App />);
  const input = getByTestId(appSearchInput);
  fireEvent.changeText(input, searchUser);

  await waitFor(
    () => {
      expect(queryByText(textID)).not.toBeNull();
      expect(queryByText(resultUser)).not.toBeNull();
      const userCard = getByTestId(userCardId + textID); // we select the right card item
      fireEvent.press(
        within(userCard).getByTestId('UserCard.CheckBox' + textID),
      );
      fireEvent.press(getByTestId(deleteButton));
    },
    {timeout: 3000},
  );

  await waitFor(
    () => {
      // We expect to have nothing here
      expect(queryByText(textID)).toBeNull();
      expect(queryByText(resultUser)).toBeNull();
    },
    {timeout: 1000},
  );
});

it('render the screen, make a search, select every users to be deleted', async () => {
  fetchMock.mockResponseOnce(JSON.stringify(fakeUserResponseWithMaxResults), {
    headers: headers,
    status: 200,
  });
  const searchUser = 'sebastien';
  const {queryByTestId, getByTestId, queryByText} = render(<App />);
  const input = getByTestId(appSearchInput);
  fireEvent.changeText(input, searchUser);

  await waitFor(
    () => {
      fireEvent.press(getByTestId(checkbox));
      expect(queryByText(`30 ${text('elementsSelected')}`)).not.toBeNull();
      fireEvent.press(getByTestId(deleteButton));
      expect(queryByTestId(flatList)).toBeNull();
      expect(queryByText(`0 ${text('elementsSelected')}`)).not.toBeNull();
      expect(queryByTestId(textNoFoundSearch)).not.toBeNull();
    },
    {timeout: 3000},
  );
});

it('render the screen, make a search, select every users to be duplicated', async () => {
  fetchMock.mockResponseOnce(JSON.stringify(fakeUserResponseWithMaxResults), {
    headers: headers,
    status: 200,
  });
  const searchUser = 'sebastien';
  const {queryByTestId, getByTestId, queryByText} = render(<App />);
  const input = getByTestId(appSearchInput);
  fireEvent.changeText(input, searchUser);

  await waitFor(
    () => {
      fireEvent.press(getByTestId(checkbox));
      expect(queryByText(`30 ${text('elementsSelected')}`)).not.toBeNull();
      fireEvent.press(getByTestId(copyButton));
      expect(queryByTestId(flatList)).not.toBeNull();
      fireEvent.press(getByTestId(checkbox));
      expect(queryByText(`60 ${text('elementsSelected')}`)).not.toBeNull();
      expect(queryByTestId(textNoFoundSearch)).toBeNull();
    },
    {timeout: 3000},
  );
});
