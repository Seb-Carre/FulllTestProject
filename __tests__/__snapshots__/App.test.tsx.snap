// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render the screen as snapShot expected 1`] = `
<RCTSafeAreaView
  style={
    {
      "flex": 1,
      "padding": 10,
    }
  }
>
  <View
    style={
      {
        "flex": 1,
        "padding": 10,
      }
    }
  >
    <View
      style={
        {
          "flex": 1,
        }
      }
    >
      <View
        style={
          {
            "alignItems": "center",
            "backgroundColor": "gray",
            "height": 64,
            "justifyContent": "center",
            "width": "100%",
          }
        }
      >
        <Text
          style={
            [
              {},
              undefined,
            ]
          }
        >
          Github Search
        </Text>
      </View>
      <TextInput
        onChangeText={[Function]}
        placeholder="Search input"
        placeholderTextColor="black"
        style={
          {
            "backgroundColor": "lightgray",
            "borderRadius": 5,
            "fontSize": 17,
            "fontWeight": "600",
            "marginBottom": 24,
            "marginLeft": 16,
            "marginRight": 16,
            "marginTop": 24,
            "padding": 20,
            "paddingLeft": 90,
          }
        }
        testID="App.SearchInput"
      />
      <View
        style={
          {
            "flexDirection": "row",
            "marginBottom": 24,
            "marginLeft": 16,
            "marginRight": 16,
            "marginTop": 24,
          }
        }
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "backgroundColor": "#44a1f8",
              "borderRadius": 4,
              "borderWidth": 4,
              "height": 20,
              "justifyContent": "center",
              "width": 20,
            }
          }
          testID="App.CheckBox"
        >
          <View
            style={{}}
            testID="Checkbox.viewStyle"
          />
        </View>
        <Text
          style={
            [
              {},
              {
                "flex": 1,
                "fontSize": 17,
                "fontWeight": "bold",
                "lineHeight": 20,
                "marginLeft": 10,
              },
            ]
          }
        >
          0
           
          <Text
            style={
              [
                {},
                {
                  "fontWeight": "normal",
                },
              ]
            }
          >
            elements selected
          </Text>
        </Text>
        <View
          style={
            {
              "flexDirection": "row",
            }
          }
        >
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            testID="App.CopyButton"
          >
            <Image
              source={
                {
                  "testUri": "../../../assets/images/copy.png",
                }
              }
              style={
                {
                  "height": 20,
                  "width": 20,
                }
              }
            />
          </View>
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            testID="App.DeleteButton"
          >
            <Image
              source={
                {
                  "testUri": "../../../assets/images/trash.png",
                }
              }
              style={
                {
                  "height": 20,
                  "width": 20,
                }
              }
            />
          </View>
        </View>
      </View>
      <View
        style={
          {
            "alignContent": "center",
            "alignItems": "center",
            "flex": 1,
            "marginLeft": 16,
            "marginRight": 16,
          }
        }
      >
        <View
          style={
            {
              "flex": 1,
            }
          }
        >
          <Text
            style={
              [
                {
                  "fontSize": 20,
                  "fontWeight": "600",
                },
                undefined,
              ]
            }
            testID="App.BeginSearch"
          >
            Please ! Begin the search by typing the user you're looking for.
          </Text>
        </View>
      </View>
    </View>
  </View>
</RCTSafeAreaView>
`;
